Smt/Term.lean

```diff
@@ -61,29 +61,36 @@ def quoteSymbol (s : String) : String :=
   else s
 
 /-- Print given `Term` in SMT-LIBv2 format. -/
-partial def toString : Term → String
+protected partial def toString : Term → String
   | literalT l    => l             -- do not quote theory literals
   | symbolT n     => quoteSymbol n
   | t@(arrowT ..) =>
     let ss := arrowToList t
-    "(" ++ String.intercalate " " (ss.init.map toString) ++ ") "
+    "(" ++ String.intercalate " " (ss.init.map Term.toString) ++ ") "
         ++ ss.getLast!.toString
   | t@(appT ..) =>
-    let ts := List.reverse <| appToList t
-    "(" ++ String.intercalate " " (ts.map toString) ++ ")"
+    let ts := appToList [] t
+    "(" ++ String.intercalate " " (ts.map Term.toString) ++ ")"
   | forallT n s t => s!"(forall (({quoteSymbol n} {s.toString})) {t.toString})"
   | existsT n s t => s!"(exists (({quoteSymbol n} {s.toString})) {t.toString})"
   | letT n t b    => s!"(let (({quoteSymbol n} {t.toString})) {b.toString})"
   where
     arrowToList : Term → List Term
       | arrowT d c => d :: arrowToList c
       | s          => [s]
-    appToList : Term → List Term
-      | appT f t => t :: appToList f
-      | s        => [s]
+    appToList (acc : List Term) : Term → List Term
+      -- We hardcode support for the `(_ extract i j)` term which needs to be parenthesized
+      -- since SMT-LIB does not curry applications. `(_ BitVec n)` does not need special
+      -- casing since it is not a function, so it should not be the head of an `appT`.
+      | appT (symbolT "_") (symbolT "extract") =>
+        let i := acc.get! 0
+        let j := acc.get! 1
+        literalT s!"(_ extract {Term.toString i} {Term.toString j})" :: acc.drop 2
+      | appT f t => appToList (t :: acc) f
+      | s        => s :: acc
```

Comment:
I think it would work, but is there an advantage to doing so? Currently our `Smt.Term`s map exactly to what is emitted as SMT-LIB, modulo parenthetization.

Reply:
It's just hard to think of `_` as a function to apply on terms. So, it feels like we're hacking the data-structure and its printer to print what we want. I am suggesting we hack only one 🙂

(I know you would rather do both properly, but I refuse to add `_` as a constructor)

Reply:
Part [3.3](https://smtlib.cs.uiowa.edu/papers/smt-lib-reference-v2.6-r2021-05-12.pdf) of the spec states "indexed identifiers are defined more systematically as the application of the reserved word _ to a symbol and one or more indices". So it does specify an "application" of ` _`. We could alternatively represent `appT (symbolT "_") (symbolT sym)` as `indexedT sym`, but I'm not sure if this really helps with anything. What do you think?

Reply:
Like I said above, I am against adding `_` as a constructor. I will ask Prof. Cesare for his opinion on this matter. For now, let's merge this PR as is.